File: javascript/binary-search-tree/binary-search-tree.js
class BinarySearchTree {

    constructor(val) {
        this.val = val;
        this.left = this.right = null;
    }

    insert(val) {
        if (val < this.val) {
            if (!this.left) {
                this.left = new BinarySearchTree(val);
            } else {
                this.left.insert(val);
            }
        } else {
            if (!this.right) {
                this.right = new BinarySearchTree(val);
            } else {
                this.right.insert(val);
            }
        }
        return this;
    }

    find(val) {
        if (val === this.val) {
            return this;
        }
        if (val < this.val) {
            if (!this.left) {
                return null;
            }
            return this.left.find(val);
        } else {
            if (!this.right) {
                return null;
            }
            return this.right.find(val);
        }
    }

    traverse(func) {
        if (this.left) {
            this.left.traverse(func);
        }

        func(this.val);

        if (this.right) {
            this.right.traverse(func)
        }
    }
}


module.exports = BinarySearchTree;

/*

const root = new BinarySearchTree(5);

root.insert(7)
    .insert(8)
    .insert(9)
    .insert(10)
    .insert(6);

root.insert(4)
    .insert(2)
    .insert(1)
    .insert(3);

console.log(root);

console.log(root.find(8));
console.log(root.find(12));
console.log(root.traverse());

*/

File: golang/binary-search-tree.go
package main

import "fmt"

type BinarySearchTree struct {
    Val string
    Left *BinarySearchTree
    Right *BinarySearchTree
}

func (T *BinarySearchTree) Insert(val string) {
    if (val < T.Val) {
        if (T.Left != nil) {
            T.Left.Insert(val)
        } else {
            T.Left = &BinarySearchTree{val, nil, nil}
        }
    } else {
        if (T.Right != nil) {
            T.Right.Insert(val)
        } else {
            T.Right = &BinarySearchTree{val, nil, nil}
        }
    }
}

func (T *BinarySearchTree) String() (string) {
    if (T == nil) {
        return "nil"
    }
    res := "{" + T.Val +
    " Left: " + T.Left.String() +
    ", Right: " + T.Right.String() +
    "}"
    return res
}


func (T *BinarySearchTree) Find(val string) *BinarySearchTree {
    if (T == nil) {
        return nil
    }
    if (T.Val == val) {
        return T
    }
    if (val < T.Val) {
        return T.Left.Find(val)
    } else {
        return T.Right.Find(val)
    }
}

func (T *BinarySearchTree) Traverse(f func(*BinarySearchTree)) {
    if (T == nil) {
        return
    }
    T.Left.Traverse(f)
    f(T)
    T.Right.Traverse(f)
}

/*
func main() {

    root := &BinarySearchTree{"5", nil, nil}
    root.Insert("2")
    root.Insert("3")
    root.Insert("1")
    root.Insert("6")
    root.Insert("8")
    root.Insert("4")
    root.Insert("7")

    fmt.Println(root)
    six := root.Find("6")
    fmt.Println(six)

    root.Traverse(func (T *BinarySearchTree) {
        fmt.Println(T.Val)
    })
}
*/
